src/UnConstrainedFPOPLog.cpp

```diff
@@ -0,0 +1,319 @@
+	  printf("=prev up cost\n");
+	  up_cost_prev->print();
+	  printf("=min more(prev up cost)\n");
+	  min_prev_cost.print();
+	  throw status;
+	}
+	min_prev_cost.set_prev_seg_end(data_i-1);
+	//NO PENALTY FOR DOWN CHANGE
+	// if(data_i==27042){
+	//   printf("computing cost data_i=%d\n", data_i);
+	//   verbose=1;
+	// }
+	down_cost->set_to_min_env_of(&min_prev_cost, down_cost_prev, verbose);
+	status = down_cost->check_min_of(&min_prev_cost, down_cost_prev);
+	if(status){
+	  printf("BAD MIN ENV CHECK data_i=%d status=%d\n", data_i, status);
+	  down_cost->set_to_min_env_of(&min_prev_cost, down_cost_prev, true);
+	  printf("=prev up cost\n");
+	  up_cost_prev->print();
+	  printf("=min more(prev up cost)\n");
+	  min_prev_cost.print();
+	  printf("=prev down cost\n");
+	  down_cost_prev->print();
+	  printf("=new down cost model\n");
+	  down_cost->print();
+	  throw status;
+	}
+      }
+      down_cost->multiply(cum_weight_prev_i);
+      down_cost->add
+	(weight_vec[data_i],
+	 -data_vec[data_i]*weight_vec[data_i],
+	 0.0);
+      down_cost->multiply(1/cum_weight_i);
+    }//if(data_i initialization else update
+    cum_weight_prev_i = cum_weight_i;
+    up_cost_prev = up_cost;
+    down_cost_prev = down_cost;
+  }
+  // Decoding the cost_model_vec, and writing to the output matrices.
+  double best_cost, best_log_mean, prev_log_mean;
+  int prev_seg_end=data_count;
+  for(int i=0; i<data_count; i++){
+    mean_vec[i] = INFINITY;
+    end_vec[i] = -2;
+  }
+  for(int i=0; i<2*data_count; i++){
+    up_cost = &cost_model_mat[i];
+    intervals_mat[i] = up_cost->piece_list.size();
+    up_cost->Minimize
+      (cost_mat+i, &best_log_mean,
+       &prev_seg_end, &prev_log_mean);
+  }
+  // last segment is down (offset N) so the second to last segment is
+  // up (offset 0).
+  int prev_seg_offset = 0;
+  down_cost = &cost_model_mat[data_count*2-1];
+  down_cost->Minimize
+    (&best_cost, &best_log_mean,
+     &prev_seg_end, &prev_log_mean);
+  mean_vec[0] = exp(best_log_mean);
+  end_vec[0] = prev_seg_end;
+  int out_i=1;
+  while(0 <= prev_seg_end){
+    // up_cost is actually either an up or down cost.
+    up_cost = &cost_model_mat[prev_seg_offset + prev_seg_end];
+    //printf("decoding out_i=%d prev_seg_end=%d prev_seg_offset=%d\n", out_i, prev_seg_end, prev_seg_offset);
+    //up_cost->print();
+    if(prev_log_mean != INFINITY){
+      //equality constraint inactive
+      best_log_mean = prev_log_mean;
+    }
+    up_cost->findMean
+      (best_log_mean, &prev_seg_end, &prev_log_mean);
+    mean_vec[out_i] = exp(best_log_mean);
+    end_vec[out_i] = prev_seg_end;
+    // change prev_seg_offset and out_i for next iteration.
+    if(prev_seg_offset==0){
+      //up_cost is actually up
+      prev_seg_offset = data_count;
+    }else{
+      //up_cost is actually down
+      prev_seg_offset = 0;
+    }
+    out_i++;
+  }//for(data_i
+}
+
+// void PeakSegFPOPall
+// (int *data_vec, double *weight_vec, int data_count,
```

Comment:
you can delete all of these comments, this is some old code that I no longer use...